xdsl/irdl/declarative_assembly_format_parser.py

```diff
@@ -241,20 +243,46 @@ def add_reserved_attrs_to_directive(self, elements: list[FormatDirective]):
                 )
                 return
 
-    def resolve_types(self) -> set[str]:
+    @dataclass(frozen=True)
+    class _OperandResultResolver(Resolver[ParsingState]):
+        idx: int
+        is_operand: bool
+        resolver: Resolver[Sequence[Attribute]]
+
+        def resolve(self, a: ParsingState) -> ResolveType:
+            if self.is_operand:
+                types = a.operand_types[self.idx]
+            else:
+                types = a.result_types[self.idx]
+            assert types is not None
+            if isinstance(types, Attribute):
+                types = (types,)
+            return self.resolver.resolve(types)
+
+    def resolve_types(self) -> dict[str, Resolver[ParsingState]]:
         """
         Find out which constraint variables can be inferred from the parsed attributes.
         """
-        seen_variables = set[str]()
+        resolvers = dict[str, Resolver[ParsingState]]()
         for i, (_, operand_def) in enumerate(self.op_def.operands):
             if self.seen_operand_types[i]:
-                seen_variables |= operand_def.constr.get_resolved_variables()
+                resolvers.update(
+                    {
+                        v: self._OperandResultResolver(i, True, r)
+                        for v, r in operand_def.constr.get_resolvers().items()
+                    }
+                )
         for i, (_, result_def) in enumerate(self.op_def.results):
             if self.seen_result_types[i]:
-                seen_variables |= result_def.constr.get_resolved_variables()
-        return seen_variables
+                resolvers.update(
+                    {
+                        v: self._OperandResultResolver(i, False, r)
+                        for v, r in result_def.constr.get_resolvers().items()
+                    }
+                )
+        return resolvers
```

Comment:
What happens if there's an overlap in `v`s? Worth a comment if that's not possible? My instinct here would be to create two dictionaries, probably with two dictionary comprehensions, and then take their union. (Maybe assert that the intersection of keys is empty)

Reply:
If there's an overlap in `v`s then the second one encountered will be used. This would be happen if a variable can be inferred from two different places in the assembly format. Which resolver is used shouldn't make any difference for any operation that successfully verifies.

Reply:
how do you feel about a resolver that asserts at runtime that the two resolved results are the same?

Reply:
I imagine this would be easy to do with the new class-based system. I'm not sure how I feel about putting what is effectively verification as a check in parsing.